pkg/uploader/provider/restic.go

```diff
@@ -13,22 +13,162 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	"github.com/vmware-tanzu/velero/pkg/uploader"
+	"github.com/vmware-tanzu/velero/pkg/util/filesystem"
 )
 
+// BackupFunc mainly used to make testing more convenient
+var ResticBackupCMDFunc = restic.BackupCommand
+var ResticRestoreCMDFunc = restic.RestoreCommand
+
+type resticProvider struct {
+	repoIdentifier  string
+	credentialsFile string
+	caCertFile      string
+	cmdEnv          []string
+	extraFlags      []string
+	backupTags      map[string]string
+	log             logrus.FieldLogger
+	bsl             *velerov1api.BackupStorageLocation
+}
+
 func NewResticUploaderProvider(
 	repoIdentifier string,
 	bsl *velerov1api.BackupStorageLocation,
 	credGetter *credentials.CredentialGetter,
 	repoKeySelector *v1.SecretKeySelector,
 	log logrus.FieldLogger,
 ) (Provider, error) {
-	return nil, nil //TODO
+	provider := resticProvider{
+		repoIdentifier: repoIdentifier,
+		log:            log,
+		bsl:            bsl,
+	}
+
+	var err error
+	provider.credentialsFile, err = credGetter.FromFile.Path(repoKeySelector)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating temp restic credentials file")
+	}
+
+	// if there's a caCert on the ObjectStorage, write it to disk so that it can be passed to restic
+	if bsl.Spec.ObjectStorage != nil && bsl.Spec.ObjectStorage.CACert != nil {
+		provider.caCertFile, err = restic.TempCACertFile(bsl.Spec.ObjectStorage.CACert, bsl.Name, filesystem.NewFileSystem())
+		if err != nil {
+			return nil, errors.Wrap(err, "error create temp cert file")
+		}
+	}
+
+	provider.cmdEnv, err = restic.CmdEnv(bsl, credGetter.FromFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "error generating repository cmnd env")
+	}
+
+	// #4820: restrieve insecureSkipTLSVerify from BSL configuration for
+	// AWS plugin. If nothing is return, that means insecureSkipTLSVerify
+	// is not enable for Restic command.
+	skipTLSRet := restic.GetInsecureSkipTLSVerifyFromBSL(bsl, log)
+	if len(skipTLSRet) > 0 {
+		provider.extraFlags = append(provider.extraFlags, skipTLSRet)
+	}
+
+	return &provider, nil
+}
+
+// Not implement yet
+func (rp *resticProvider) Cancel() {
+}
+
+func (rp *resticProvider) Close(ctx context.Context) error {
+	if err := os.Remove(rp.credentialsFile); err != nil {
+		return errors.Wrapf(err, "failed to remove file %s", rp.credentialsFile)
+	}
+	if err := os.Remove(rp.caCertFile); err != nil {
+		return errors.Wrapf(err, "failed to remove file %s", rp.caCertFile)
+	}
+	return nil
+}
+
+// RunBackup runs a `backup` command and watches the output to provide
+// progress updates to the caller.
+func (rp *resticProvider) RunBackup(
+	ctx context.Context,
+	path string,
+	tags map[string]string,
+	parentSnapshot string,
+	updater uploader.ProgressUpdater) (string, error) {
+	if updater == nil {
+		return "", errors.New("Need to initial backup progress updater first")
+	}
+
+	log := rp.log.WithFields(logrus.Fields{
+		"path":           path,
+		"parentSnapshot": parentSnapshot,
+	})
+
+	rp.backupTags = tags
```

Comment:
The `backupTags` will be overwritten if call `RunBackup` in paralell

Reply:
yeah, I've removed the backupTags by passing  tags directly into `GetSnapshotCommand`